tslearn/nonmyopic.py

```diff
@@ -0,0 +1,370 @@
+                            ],
+                            [
+                                self.clusters_["ck_cm{0}".format(k)][-1][
+                                    1
+                                ][0]
+                                / column_sum[1],
+                                self.clusters_["ck_cm{0}".format(k)][-1][
+                                    1
+                                ][1]
+                                / column_sum[1],
+                            ],
+                        ]
+                    )
+                    self.pyhatyck_["pyhatycks{0}".format(k)].append(
+                        matrix_proba_current_cluster_class
+                    )
+
+    def get_cluster_probas(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        distances_clusters = []
+        centroids = self.cluster_.cluster_centers_[:, :len(X)]
+        for k in range(0, self.n_clusters):
+            euclidean_dist = np.linalg.norm(
+                X - centroids[k]
+            )
+            distances_clusters.append(euclidean_dist)
+        distances_clusters = np.asarray(distances_clusters)
+        minimum_distance = np.min(distances_clusters)
+        average_distance = np.mean(distances_clusters)
+        minimum_distance_to_average = (
+                                              average_distance - minimum_distance
+                                      ) / average_distance
+        sum_sk = 0
+        sk = []
+        for k in range(0, self.n_clusters):
+            d_k = distances_clusters[k]
+            Delta_k = (average_distance - d_k) / average_distance
+            proba_current_cluster = 1 / (
+                np.exp(self.lamb * minimum_distance_to_average)
+                + np.exp(self.lamb * (minimum_distance_to_average - Delta_k))
+                + 10e-6
+            )
+            sk.append(proba_current_cluster)
+            sum_sk = sum_sk + sk[-1]
+        final_sk = [(x / sum_sk) for x in sk]
+        return final_sk
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        cost = 0
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[-1]
+        for k in range(0, self.n_clusters):
```

Comment:
The formula is a sum of sums. Probably if your data was stored in `numpy` arrays as it should, this function could be vectorized.